File: usig/sgx/test/usig_test.c
// Copyright (c) 2018 NEC Laboratories Europe GmbH.
//
// Authors: Sergey Fedorov <sergey.fedorov@neclab.eu>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#undef NDEBUG // make sure `assert()` is not an empty macro
#include <assert.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <stdio.h>

#include "usig.h"

const char *enclave_file;

// NOTE: We only do *limited* testing here. For instance, we don't
// check correctness of the signature (certificate) produced by the
// enclave because it's cumbersome to do so in C. That's easier to
// test in Golang, together with the certificate verification
// implementation.

static void test_init_destroy()
{
        sgx_enclave_id_t eid;

        assert(usig_init(enclave_file, &eid, NULL, 0) == SGX_SUCCESS);
        assert(usig_destroy(eid) == SGX_SUCCESS);
}

static inline bool signature_is_equal(usig_ui *ui1, usig_ui *ui2)
{
        return memcmp(&ui1->signature, &ui2->signature, sizeof(ui1->signature)) == 0;
}

static void test_seal_key()
{
        sgx_enclave_id_t usig;
        void *sealed_data;
        size_t sealed_data_size;

        assert(usig_init(enclave_file, &usig, NULL, 0) == SGX_SUCCESS);
        assert(usig_seal_key(usig, &sealed_data,
                             &sealed_data_size) == SGX_SUCCESS);
        assert(usig_destroy(usig) == SGX_SUCCESS);
        assert(usig_init(enclave_file, &usig, sealed_data,
                         sealed_data_size) == SGX_SUCCESS);
        free(sealed_data);
        assert(usig_destroy(usig) == SGX_SUCCESS);
}

static void test_create_ui()
{
        sgx_enclave_id_t usig;
        void *sealed_data;
        size_t sealed_data_size;
        usig_ui ui1, ui2, ui3;
        sgx_sha256_hash_t digest = "TEST DIGEST";

        assert(usig_init(enclave_file, &usig, NULL, 0) == SGX_SUCCESS);
        assert(usig_seal_key(usig, &sealed_data,
                             &sealed_data_size) == SGX_SUCCESS);
        assert(usig_create_ui(usig, digest, &ui1) == SGX_SUCCESS);
        // The first counter value must be zero
        assert(ui1.counter == 1);
        assert(usig_create_ui(usig, digest, &ui2) == SGX_SUCCESS);
        // The counter must be monotonic and sequential
        assert(ui2.counter == ui1.counter + 1);
        // Certificate must be unique for each counter value
        assert(!signature_is_equal(&ui1, &ui2));
        // Epoch should be the same for the same USIG instance
        assert(ui1.epoch == ui2.epoch);
        assert(usig_destroy(usig) == SGX_SUCCESS);
        // Recreate USIG using the sealed secret from the first instance
        assert(usig_init(enclave_file, &usig, sealed_data,
                         sealed_data_size) == SGX_SUCCESS);
        assert(usig_create_ui(usig, digest, &ui3) == SGX_SUCCESS);
        // Must fetch a fresh counter value
        assert(ui3.counter == 1);
#ifndef SGX_SIM_MODE
        // Apparently, SGX SDK in the simulation mode uses current
        // time in *seconds* to seed random number generation. We
        // don't want to wait that long and skip these checks.
        // Check for uniqueness of the epoch and certificate produced
        // by the new instance of the enclave
        assert(ui1.epoch != ui3.epoch);
        assert(!signature_is_equal(&ui1, &ui3));
#endif
        assert(usig_destroy(usig) == SGX_SUCCESS);
        free(sealed_data);
}

int main(int argc, const char **argv)
{
        assert(argc == 2);
        enclave_file = argv[1];

        test_init_destroy();
        test_seal_key();
        test_create_ui();

        puts("PASS");
}

File: usig/usig_test.go
// Copyright (c) 2018 NEC Laboratories Europe GmbH.
//
// Authors: Wenting Li     <wenting.li@neclab.eu>
//          Sergey Fedorov <sergey.fedorov@neclab.eu>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package usig_test

import (
	"testing"

	"github.com/golang/mock/gomock"
	"github.com/stretchr/testify/assert"

	"github.com/nec-blockchain/minbft/usig"
	mock_usig "github.com/nec-blockchain/minbft/usig/mocks"
)

func TestUIMarshalUnmarshal(t *testing.T) {
	uiIn := &usig.UI{
		Epoch:   13,
		Counter: 42,
		Cert:    []byte("test cert"),
	}
	bytes, err := uiIn.MarshalBinary()
	assert.NoError(t, err)
	uiOut := new(usig.UI)
	err = uiOut.UnmarshalBinary(bytes)
	assert.NoError(t, err)
	assert.True(t, assert.ObjectsAreEqual(uiIn, uiOut),
		"Unmarshaled UI doesn't equal marshaled")
}

// TestMockUsig shows the usage of mocked USIG
func TestMockUsig(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockUsig := mock_usig.NewMockUSIG(ctrl)

	uis := []usig.UI{
		{Counter: 1},
		{Counter: 2},
		{Counter: 3},
	}
	uisExp := []usig.UI{
		{Counter: 1},
		{Counter: 2},
		{Counter: 3},
	}
	gomock.InOrder(
		mockUsig.EXPECT().CreateUI(gomock.Any()).Return(&uis[0], nil),
		mockUsig.EXPECT().CreateUI(gomock.Any()).Return(&uis[1], nil),
		mockUsig.EXPECT().CreateUI(gomock.Any()).Return(&uis[2], nil),
	)
	for i := range uis {
		mockUsig.EXPECT().VerifyUI(gomock.Any(), &uis[i], gomock.Any()).Return(nil).AnyTimes()
	}

	testMsg := []byte("any message")
	ui0, err := mockUsig.CreateUI(testMsg)
	assert.NoError(t, err)
	assert.True(t, assert.ObjectsAreEqual(&uisExp[0], ui0))
	ui1, err := mockUsig.CreateUI(testMsg)
	assert.NoError(t, err)
	assert.True(t, assert.ObjectsAreEqual(&uisExp[1], ui1))
	ui2, err := mockUsig.CreateUI(testMsg)
	assert.NoError(t, err)
	assert.True(t, assert.ObjectsAreEqual(&uisExp[2], ui2))
	assert.NoError(t, mockUsig.VerifyUI(testMsg, &uis[0], []byte("usig id")))
}
